### ISSUE TITLE
Use x360ce as a plugin

### ISSUE THREAD

**Issue by AndersMalmgren** [x360ce/x360ce#191]
Hi, I would love to incorporate this in FreePIE https://github.com/AndersMalmgren/FreePIE

Have you guys have any thoughts of creating a API for it so that you can load x360ce as an assembly from a third party program?

**Comment by Squall-Leonhart** [comment]
This question has been posed before,

The answer then was "No" if i remember right, assigning tapcio to see if this has changed.

**Comment by AndersMalmgren** [comment]
Right now it is working using FreePIE > vJoy > x360ce. But it would nice to skip the vjoy step and communicate directly with the x360ce driver

**Comment by Squall-Leonhart** [comment]
x360ce is not a driver

anyone is free to take x360ce and incorporate its emulation features into any application they want.

**Comment by Nucleoprotein** [comment]
I do not know what you want really. x360ce is NOT a driver. Is only MS XInput 1.x to DirectInput wrapper. ie. it gets input from DirectInput and returns it as XInput API. Usage in third party program is not different that normal MS XInput API.

PS: x360ce uses some additional features to make it work in games using API hooks, it's named InputHook. I can expose InputHook API trough DLL exports if this is required. We already expose `Reset` for reloading settings.

**Comment by AndersMalmgren** [comment]
So there is a way of letting third party software send commands directly to the emulation and skip directinput?

**Comment by Nucleoprotein** [comment]
What ? DirectInput is API -from- where x360ce -gets- input data, ie. x360ce expose XInput API to game and gets input from controller using DirectInput API. Example: 
Game calls XInputGetState, x360ce calls IDirectInputDevice8::Poll then IDirectInputDevice8::GetDeviceState to get data from device. Next, it translates this data to XInput compatible data and return it to caller (game).

Sending data directly to emulation ? For what ? You can send it to real MS XInput API instead - it same.

**Comment by AndersMalmgren** [comment]
Yes but it requires you to install vjoy if your input device is not  a directinput device, like a oculus rift HMD or a Wiimote

**Comment by Squall-Leonhart** [comment]
Closing because this is never going to happen.

**Comment by PsykomantaBrain** [comment]
I would add my vote to this feature. FreePIE being able to talk directly to x360ce would streamline the process of creating custom control schemes a lot. 

Currently the only solution is to set up FreePIE to drive a virtual joystick device and map that into x360ce. A direct FreePIE->x360ce interface would be incredibly useful.

Cheers

**Comment by Nucleoprotein** [comment]
I don't know what you want, because FreePIE can use DirectInput and Xinput devices. So what you really want ? 
Maybe you want to use FreePIE as API on top of x360ce, ie. using this: https://github.com/AndersMalmgren/FreePIE/wiki/IO-Plugin in xinput*.dll wrappers ?

**Comment by AndersMalmgren** [comment]
I think you guys are missing what we want. Today we need the extra step of Vjoy so that we can feed x360xe with the input. We want to skip this step by letting FreePIE communicate withg x360xe directly. For example through a shared memory pipeline?

edit: PsykomantaBrains explanation below is in more detail

**Comment by PsykomantaBrain** [comment]
I can't say for sure what's needed, I'm not the dev for FreePIE, but the idea is exposing controls so apps like FreePIE are able to feed input directly to the x360ce-emulated controllers. 

For instance, right now, from FreePIE, we have to do this:

`vJoyDevice.someAxis = myCustomDevice.anotherAxis`

And then, in x360ce's config, we map that someAxis to the controller that was assigned for vJoy.

The idea would be to skip vJoy altogether, by adding a way to have FreePIE directly access (write to) the bindings of x360ce controllers.

Thanks for the quick reply! 

EDIT: Ninja'd!

Cheers

**Comment by AndersMalmgren** [comment]
In other words in freepie you would then do

``` python
xinput.x = someOtherDevice.x
```

And then FreePIE and x360ce will deliver that emulated value to the game.

edit: I think shared memory would work perfect, if freepie writes to the shared memory and your spoofed dll picks this up it could ignore config.ini all together and listen to whats on the shared memory feed instead, I think it would be very little change in your code?

**Comment by EJocys** [comment]
Just to clear things out. I will describe very simple and limited way to feed data into XInput and maybe we could go with it further. At the moment, XInput allows to read controller state where State type is struct:

```
ErrorCode XInputGetState(int dwUserIndex, out State pState);
```

Lets say we will create extra options inside INI file:

```
[Options]
Pad1EmulationSource=1
Pad2EmulationSource=0
Pad3EmulationSource=0
Pad4EmulationSource=0
```

and extra method inside XInput DLL:

```
ErrorCode XInputSetState(int dwUserIndex, State pState)
{
    EmulatedValues[dwUserIndex] = pState;
}
```

XInputGetState will be extended like this:

```
ErrorCode XInputGetState(int dwUserIndex, out State pState)
{
    if (Pad1EmulationSource[dwUserIndex] == 1)
    {
        pState = EmulatedValues[dwUserIndex];
        return 0;
    }
    // use DInput and non emulated sources...
}
```

Now there is an issue. I am not an expert in this, but I guess that third party program have to use some inter-process communication magic, find XInput DLL inside the memory and call XInputSetState method in order to feed data into it.

This means that we must create extra functions in order to collect emulated PAD State from other places (place in the memory or use inter-process communication to retrieve State values from other process):

```
ErrorCode XInputGetState(int dwUserIndex, out State pState)
{
    if (Pad1EmulationSource[dwUserIndex] == 1)
    {
        pState = EmulatedValues[dwUserIndex];
        return 0;
    }
    else if (Pad1EmulationSource[dwUserIndex] == 2)
    {
        // Get State structure from specified memory address.
    }
    else if (Pad1EmulationSource[dwUserIndex] == 3)
    {
        // Get State structure from other process by calling its method.
        // Getting data directly from the feeder probably would be the best and fastest,
        // because there is no delay created between writing and reading.
        // and feeder can tailor every State depending on read time and read count.
    }
    else if (Pad1EmulationSource[dwUserIndex] == 4)
    {
        // Some other method.
    }
    // use DInput and non emulated sources...
}
```

**Comment by Nucleoprotein** [comment]
XInputSetState name is already taken (vibration).

**Comment by EJocys** [comment]
@tapcio Correct. We can pick different name, but I guess same name will be fine too, because our method will have different signature from XInputSetState(int dwUserIndex, ref Vibration pVibration);

**Comment by AndersMalmgren** [comment]
I think you would be best off abstracting input to an interface and then have one DirectInput implementation and one SharedMemoryInput implementation of said interface. That way you can add any number of input sources without changing a line of code. 

Also the ini file would be useless for SharedMemoryInput  since we can do the config in the input source program (For example FreePIE).

**Comment by Squall-Leonhart** [comment]
i don't think we should get involved in this at all, the author of  glovepie should implement their own xinput emulation.

**Comment by AndersMalmgren** [comment]
Why not improve on an already excellent product?

**Comment by AndersMalmgren** [comment]
"_Now there is an issue. I am not an expert in this, but I guess that third party program have to use some inter-process communication magic, find XInput DLL inside the memory and call XInputSetState method in order to feed data into it._"

With shared memory there is no need to have a dll inside memory. We will share a memory pipeline. FreePIE will write a struct there for you to read. Or rather a array of structs since you can emulate more than one Xinput. 

edit: You can look here how we use it in our fake TrackIr dll https://github.com/AndersMalmgren/FreePIE/blob/master/Lib/TrackIR/NPClient/shared_memory.h

**Comment by Nucleoprotein** [comment]
But x360ce is a dll ... x360ce.exe is **only** configuration app.

**Comment by AndersMalmgren** [comment]
Yes hosted by the XInput driven app (Game etc). But that does not change anything. It will still be hosted the same. The difference is just how you get input. Keep in mind I do not have 100% insight in the x360ce code so I can be wrong. But my guess is you read the state of DirectInput each time the game checks state?

**Comment by Nucleoprotein** [comment]
Shared memory is out of question - this will never happen, but we can add dll export to feed x360ce with data instead of getting it from DirectInput.

**Comment by AndersMalmgren** [comment]
But to get the data into the xinput dll instance hosted by the game you need to use somekind of transport? Shared memory is the way you usually do it

edit: spelling

**Comment by AndersMalmgren** [comment]
My idea is this.

FreePIE > shared memory > faked xinput.dll reads shared memory > game reads fake xinput from faked xinput dll

edit: Only difference from how it works today is,

FreePIE > DirectInput > faked xinput.dll reads DirectInput > game reads fake xinput from faked xinput dll

**Comment by Nucleoprotein** [comment]
We are using DirectInput8 to get device state.
Shared memory will **never** happen.

**Comment by AndersMalmgren** [comment]
see edit please

**Comment by AndersMalmgren** [comment]
Then what are we discussing?

**Comment by Nucleoprotein** [comment]
Then why use shared memory ?
You can do:
FreePIE > calls x360ceSetState (or whatever we can call it) in xinput dll > game reads data

**Comment by AndersMalmgren** [comment]
But the dll hosted by FreePIE and hosted by the game will be two different instances running in two different app domains

**Comment by Nucleoprotein** [comment]
FreePIE is not attaching to process/injecting etc ?

**Comment by AndersMalmgren** [comment]
FreePIE is a stand alone application. Sometimes we inject ourself in similar manner with faked API dll
But in those cases the main FreePIE process communicates using a shared memory pipeline with the fake dll

**Comment by AndersMalmgren** [comment]
Here is a good lightweight example when we fake/inject ourself into a third party process (Game)

https://github.com/AndersMalmgren/FreePIE/blob/master/Lib/Sixense/Emulation/Code/src/sixense_fake.cpp

**Comment by Nucleoprotein** [comment]
I reopened this, but this must wait until next 'milestone' update.

**Comment by EJocys** [comment]
There is a suggestion I've implemented on one of my projects. Feeding data into application by using simple UDP protocol. On C#, you can send and receive data in less than ~15 lines. For example. X360CE DLL could start simple UDP listener on 127.0.0.1:43500, then, any application could feed data into X360CE by simply sending serialized PAD state. C# example on how to send message to X360CE DLL:

```
var pState = ...
var clientSock = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
var address = IPAddress.Parse("127.0.0.1");
var remoteEP = new IPEndPoint(address, 43500);
clientSock.Connect(remoteEP);
var formatter = new BinaryFormatter();
var stream = new MemoryStream();
formatter.Serialize(stream, pState);
var bytes = stream.ToArray();
var bytes = System.Text.Encoding.UTF8.GetBytes(message);
clientSock.Send(bytes);
clientSock.Close();
```

X360CE will get UDP message. Convert bytes back to struct and and set:

```
 EmulatedValues[dwUserIndex] = pState;
```

Options inside INI could look like this

```
[Options]
' Start UDP Server - for Receiving data.
UdpServerEnable=1
UdpServerAddress=127.0.0.1
UdpServerPort=43500
' Start UDP Client - Sends data every time XInputGetState is called.
UdpClientEnable=1
UdpClientAddress=192.168.0.1
UdpClientPort=43500
```

At least on C#, implementation is quite simple. Not sure about C++. Side effect of this - you can use your controller to play game on PC which is located on the other end of the World :). If game supports spectator mode (like Battlefield), then this is like remote game desktop - you could take control of your friend's character :).

**Comment by AndersMalmgren** [comment]
If you had an interface input could be anything. For me Shared mem is better since it has very little latency
